common/servicecacher/cacher.go

```diff
@@ -2,65 +2,60 @@ package servicecacher
 
 import (
 	"sync"
-	"sync/atomic"
 
 	"github.com/cocaine/cocaine-framework-go/cocaine"
 )
 
-type Cacher interface {
-	Get(name string) (Service, error)
+func NewService(n string, a ...interface{}) (Service, error) {
+	return cocaine.NewService(n, a...)
 }
 
 type Service interface {
 	Call(name string, args ...interface{}) chan cocaine.ServiceResult
 	Close()
 }
 
-type cache map[string]Service
+type ServiceBurner func(string, ...interface{}) (Service, error)
+
+// cacher
+type Cacher interface {
+	Get(string, ...interface{}) (Service, error)
+}
 
 type cacher struct {
 	mutex sync.Mutex
-	data  atomic.Value
+	fun   ServiceBurner
+	cache map[string]*entry
 }
 
-func NewCacher() Cacher {
-	c := &cacher{}
-	c.data.Store(make(cache))
-
-	return c
+func NewCacher(f ServiceBurner) Cacher {
+	return &cacher{fun: f, cache: make(map[string]*entry)}
 }
 
-func (c *cacher) Get(name string) (s Service, err error) {
-	s, ok := c.get(name)
-	if ok {
-		return
-	}
-
-	s, err = c.create(name)
-	return
+type entry struct {
+	service Service
+	err     error
+	ready   chan struct{} // closed when res is ready
 }
 
-func (c *cacher) get(name string) (s Service, ok bool) {
-	s, ok = c.data.Load().(cache)[name]
-	return
-}
-
-func (c *cacher) create(name string) (Service, error) {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-	s, ok := c.get(name)
-	if ok {
-		return s, nil
+func (c *cacher) Get(name string, args ...interface{}) (Service, error) {
+	var endpoint string
+	if len(args) == 1 {
+		endpoint, _ = args[0].(string)
 	}
+	key := name + endpoint
 
-	s, err := cocaine.NewService(name)
-	if err != nil {
-		return nil, err
+	c.mutex.Lock()
+	e := c.cache[key]
+	if e == nil || e.err != nil { // first request or service error
+		e = &entry{ready: make(chan struct{})}
+		c.cache[key] = e
+		c.mutex.Unlock()
+		e.service, e.err = c.fun(name, args...)
```

Comment:
e.err must be set under a lock, because a concurent Get can lead to reading a value cached by key. Look at 51